src/components/js-tracer/ExpensiveChart.js

```diff
@@ -0,0 +1,167 @@
+import * as JsTracer from '../../profile-logic/js-tracer';
+
+import type { JsTracerTable } from '../../types/profile';
+import type { JsTracerTiming } from '../../types/profile-derived';
+import type {
+  Milliseconds,
+  UnitIntervalOfProfileRange,
+  CssPixels,
+} from '../../types/units';
+import type { PreviewSelection } from '../../types/actions';
+import type {
+  ExplicitConnectOptions,
+  ConnectedProps,
+} from '../../utils/connect';
+
+require('./index.css');
+
+const ROW_HEIGHT: CssPixels = 16;
+
+type OwnProps = {|
+  // This component can only be mounted if the JsTracerTable exists in the state.
+  +jsTracerTable: JsTracerTable,
+  +showJsTracerSummary: boolean,
+|};
+
+type DispatchProps = {|
+  +updatePreviewSelection: typeof updatePreviewSelection,
+|};
+
+type StateProps = {|
+  +timeRange: { start: Milliseconds, end: Milliseconds },
+  +interval: Milliseconds,
+  +threadIndex: number,
+  +previewSelection: PreviewSelection,
+|};
+
+type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;
+
+/**
+ * This component uses the timing information for JS tracer events. This information
+ * can be quite expensive to compute. This component is responsible for encapsulating
+ * the complexity of deferring that calculation until the proper moment in the
+ * component life cycle.
+ */
+class JsTracerExpensiveChart extends React.PureComponent<Props> {
+  /**
+   * Determine the maximum zoom of the viewport.
+   */
+  getMaximumZoom(): UnitIntervalOfProfileRange {
+    const { timeRange: { start, end }, interval } = this.props;
+    return interval / (end - start);
+  }
+
+  render() {
+    const {
+      timeRange,
+      threadIndex,
+      jsTracerTable,
+      previewSelection,
+      updatePreviewSelection,
+      showJsTracerSummary,
+    } = this.props;
+
+    const jsTracerTimingRows = _getJsTracerTimingRows(
+      jsTracerTable,
+      showJsTracerSummary
+    );
+
+    // The viewport needs to know about the height of what it's drawing, calculate
+    // that here at the top level component.
+    const maxViewportHeight = jsTracerTimingRows.length * ROW_HEIGHT;
+
+    return (
+      <JsTracerCanvas
+        key={threadIndex}
+        viewportProps={{
+          timeRange,
+          previewSelection,
+          maxViewportHeight,
+          viewportNeedsUpdate,
+          maximumZoom: this.getMaximumZoom(),
+          marginLeft: TIMELINE_MARGIN_LEFT,
+          marginRight: TIMELINE_MARGIN_RIGHT,
+        }}
+        chartProps={{
+          jsTracerTimingRows,
+          jsTracerTable,
+          updatePreviewSelection,
+          rangeStart: timeRange.start,
+          rangeEnd: timeRange.end,
+          rowHeight: ROW_HEIGHT,
+          threadIndex,
+        }}
+      />
+    );
+  }
+}
+
+type JsTracerTimingWeakmap = WeakMap<JsTracerTable, JsTracerTiming[]>;
+const _jsTracerTimingWeakmap: JsTracerTimingWeakmap = new WeakMap();
+const _jsTracerLeafTimingWeakmap: JsTracerTimingWeakmap = new WeakMap();
```

Comment:
I wonder why a WeakMap and not a Map? From what I understand, we can have one table for each track, and it won't ever change, so will never be removed from the WeakMap. WeakMaps come at a cost so I think we should use a Map instead.

Reply:
Resolved by changing to a keyed component.